File: src/snippets/guards/quick-start/vercel-ai/agent.ts
import { launchArcjet, policyInput } from "@arcjet/guard";
import { rampart } from "@arcjet/sensitive-info-rampart";
import {
  aiToolsContext,
  createAgentContext,
  guardTool,
} from "@arcjet/guard/vercel-ai/v7";
import { generateText, stepCountIs, tool } from "ai";
import { z } from "zod";

// Placeholder for your mail transport.
const emailProvider = {
  send: async (_: { to: string; body: string }) => ({ ok: true }),
};

// Create one Arcjet client and reuse it across agent runs. Rampart
// evaluates the policy's LOCAL inputs on this machine, so the email body
// never leaves your application.
const arcjet = launchArcjet({
  key: process.env.ARCJET_KEY!,
  sensitiveInfoBackend: rampart(),
});

export async function runEmailAgent(
  user: {
    id: string;
    allowedRecipients: string[];
    record: {
      name: string;
      bankAccount: string;
      routingNumber: string;
    };
  },
  prompt: string,
) {
  // This read-only tool gives the model the current customer's
  // account data.
  const getClientRecord = tool({
    description: "Get the account details on file for the current customer",
    inputSchema: z.object({}),
    execute: () => user.record,
  });

  // guardTool checks policy before the email provider can run.
  const sendEmail = guardTool(
    arcjet,
    tool({
      description: "Send an email",
      inputSchema: z.object({
        recipient: z.string().email(),
        body: z.string(),
      }),
      execute: ({ recipient, body }) =>
        emailProvider.send({ to: recipient, body }),
    }),
    {
      // The action selects the remote policy configured in step 1.
      action: "email.sent",
      // Actor and the allow list come from trusted application
      // state.
      actor: user.id,
      // Map only the values the remote policy needs.
      inputs: ({ recipient, body }) => ({
        recipient: policyInput.server.string(recipient),
        allowed_recipients: policyInput.server.stringList(
          user.allowedRecipients,
        ),
        body: policyInput.local.string(body),
      }),
    },
  );

  const tools = { getClientRecord, sendEmail };
  const context = createAgentContext();

  // The model can call either tool, but sendEmail always passes
  // through Arcjet.
  return generateText({
    model: "openai/gpt-4o-mini",
    system:
      "You are a support desk assistant. Use getClientRecord when the " +
      "request needs account details. Use sendEmail exactly once to " +
      "complete the request. Never ask a follow-up question. Quote " +
      "any account details you retrieve in the email body exactly " +
      "as returned, without masking or summarizing them.",
    prompt,
    tools,
    toolsContext: aiToolsContext(context, tools),
    stopWhen: stepCountIs(3),
  });
}

File: src/snippets/guards/quick-start/agent-framework-go/agent.go
package main

import (
	"context"
	"encoding/json"
	"os"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/arcjet/arcjet-go"
	"github.com/arcjet/arcjet-go/agentframework"
	"github.com/arcjet/arcjet-go/sensitiveinfo/rampart"
	"github.com/microsoft/agent-framework-go/agent"
	"github.com/microsoft/agent-framework-go/provider/anthropicprovider"
	"github.com/microsoft/agent-framework-go/tool"
	"github.com/microsoft/agent-framework-go/tool/functool"
)

var guard = must(arcjet.NewGuardClient(arcjet.GuardConfig{
	SensitiveInfoBackend: must(rampart.New(rampart.Options{})),
}))

type user struct {
	id                string
	allowedRecipients []string
	record            map[string]string
}

type emailArgs struct {
	Recipient string `json:"recipient"`
	Body      string `json:"body"`
}

func runEmailAgent(ctx context.Context, u user, prompt string) (string, error) {
	getClientRecord := functool.MustNew(
		functool.Config{
			Name:        "get_client_record",
			Description: "Get the account details on file for the current customer.",
		},
		func(context.Context, struct{}) (map[string]string, error) {
			return u.record, nil
		},
	)
	sendEmail := functool.MustNew(
		functool.Config{Name: "send_email", Description: "Send an email."},
		func(_ context.Context, in emailArgs) (string, error) {
			return "sent", nil
		},
	)

	tools, err := agentframework.GuardTools(
		guard,
		[]tool.Tool{getClientRecord, sendEmail},
		func(t tool.Tool) (agentframework.ToolPolicy, bool) {
			if t.Name() != "send_email" {
				return agentframework.ToolPolicy{}, false
			}
			return agentframework.ToolPolicy{
				Action: "email.sent",
				Actor: func(context.Context, json.RawMessage) (string, error) {
					return u.id, nil
				},
				Inputs: func(_ context.Context, raw json.RawMessage) (
					map[string]arcjet.GuardPolicyInput, error,
				) {
					var in emailArgs
					if err := json.Unmarshal(raw, &in); err != nil {
						return nil, err
					}
					return map[string]arcjet.GuardPolicyInput{
						"recipient": arcjet.GuardPolicyServerString(
							in.Recipient,
						),
						"allowed_recipients": arcjet.GuardPolicyServerStringList(
							u.allowedRecipients,
						),
						"body": arcjet.GuardPolicyLocalString(in.Body),
					}, nil
				},
			}, true
		},
	)
	if err != nil {
		return "", err
	}

	a := anthropicprovider.NewAgent(anthropic.NewClient(), anthropicprovider.AgentConfig{
		Model: os.Getenv("ANTHROPIC_MODEL"),
		Instructions: "You are a support desk assistant. Use " +
			"get_client_record when the request needs account details. " +
			"Use send_email exactly once to complete the request. Don't " +
			"ask a follow-up question. Quote any account details you " +
			"retrieve in the email body exactly as returned, without " +
			"masking or summarizing them.",
		Config: agent.Config{Tools: tools},
	})

	ctx = arcjet.ContextWithCorrelationID(ctx, u.id)
	resp, err := a.RunText(ctx, prompt).Collect()
	if err != nil {
		return "", err
	}
	return resp.String(), nil
}

func must[T any](value T, err error) T {
	if err != nil {
		panic(err)
	}
	return value
}
